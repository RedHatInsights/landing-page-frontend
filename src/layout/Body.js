import React, { useState } from 'react';
import {
    PageSection,
    Tabs, Tab
} from '@patternfly/react-core';

import Hero from './Hero';
import BannerCard from '../components/BannerCard';
import FancyLink from '../components/FancyLink';

import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import './Body.scss';

export function Body({ technologies }) {

    const [ activeTabKey, setActiveTabKey ] = useState(0);

    const handleTabClick = (event, tabIndex) => {
        setActiveTabKey(tabIndex);
    };

    return (
        <React.Fragment>
            <Hero
                title='Scale the management and operations of your IT infrastructure'
                subtitle='Discover purpose-built services exclusively for Red Hat subscribers'
                className='ins-c-hero__small'
                data-ouia-component-type='hero'/>
            <PageSection className='pf-m-fill ins-p-landing__content'>
                <Tabs data-ouia-navigation='true' activeKey={ activeTabKey } isSecondary onSelect={ handleTabClick }>
                    { technologies.map(({ title, id, description, image, apps }, key) => ( // map categories
                        <Tab key={ key }
                            data-ouia-component-type='tab'
                            data-ouia-component-id={ `nav-tab-${id}` }
                            data-ouia-navigation-name={ `Tab ${id}` }
                            eventKey={ key || 0 }
                            title={ <BannerCard title={ title } category-id={ id } description={ description } image={ image }/> }>
                            <div className='ins-l-app-grid'>
                                { apps.map(({ name, url, id, disabled }, key) => (<React.Fragment key={ key }>
                                    { !disabled && <FancyLink
                                        className='ins-l-app-grid__item'
                                        data-ouia-component-type='app-link'
                                        data-ouia-component-id={ `app-link-${id}` }
                                        application-id={ id }
                                        to={ `${window.location.origin}${window.insights.chrome.isBeta() ? '/beta/' : '/'}${url}` }
                                        key={ key }>
                                        { name }
                                    </FancyLink> }
                                </React.Fragment>
                                )) }
                            </div>
                        </Tab>
                    )) }
                </Tabs>
            </PageSection>
        </React.Fragment>
    );
}

Body.propTypes = {
    technologies: PropTypes.arrayOf(PropTypes.shape({
        title: PropTypes.string,
        description: PropTypes.string,
        apps: PropTypes.arrayOf(PropTypes.shape({
            id: PropTypes.string,
            name: PropTypes.string,
            url: PropTypes.string
        }))
    }))
};

Body.defaultProps = {
    technologies: []
};

function mapStateToProps({ technologies } = { technologies: { activeTechnologies: []}}) {
    return {
        technologies: technologies && technologies.activeTechnologies.filter(({ disabled }) => !disabled)
    };
}

export { mapStateToProps };
export default connect(mapStateToProps)(Body);
