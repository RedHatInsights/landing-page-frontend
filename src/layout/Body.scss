// Body.scss
.ins-p-landing__content[landing-page-type="authenticated"] {
    width: 100%;
    margin-left: auto;
    margin-right: auto;

    @media screen and (min-width: 1200px) {
        width: 80%;
    }

    .pf-c-button.pf-m-link.pf-m-inline {
        text-align: left;
    }

    [application-id="insights"] .ins-c-application-info__content-applist {
        display: grid;
        grid-template-columns: auto;
        grid-template-columns: auto auto;
        gap: 10px;
        a { margin-top: 0; }
    }
}

