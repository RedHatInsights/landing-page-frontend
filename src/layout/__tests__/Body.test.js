/*global describe, test, expect */
import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import toJson from 'enzyme-to-json';
import { mapStateToProps } from '../Body';
import ConnectedBody, { Body } from '../Body';
import { ChartSpikeIcon } from '@patternfly/react-icons';

const activeTechnologiesMock = [
    {
        title: 'title',
        id: '1',
        description: 'this is a description',
        icon: ChartSpikeIcon,
        apps: [
            {
                id: 'app-1',
                name: 'app-name-1',
                url: 'app-url-1'
            }
        ]
    },
    {
        title: 'secondtitle',
        id: '2',
        description: 'this is a second description',
        icon: ChartSpikeIcon,
        apps: [
            {
                id: 'app-2',
                name: 'app-name-2',
                url: 'app-url-2'
            }
        ]
    }
];

const mockStore = configureMockStore();
const store = mockStore(activeTechnologiesMock);

function getInput(obj) {
    return {
        technologies: {
            activeTechnologies: [ obj ]
        }
    };
}

function getOutput(obj) {
    return { technologies: [ obj ]};
}

describe('mapStateToProps', () => {
    test('should return the re-wrapped data', () => {
        const data = { foo: 'bar' };
        const output = mapStateToProps(getInput(data));
        expect(output).toEqual(getOutput(data));
    });

    test('should filter out inactive', () => {
        const data = {
            foo: 'bar',
            baz: { disabled: true }
        };
        const output = mapStateToProps(getInput(data));
        expect(output).toEqual(getOutput(data));
    });

    test('should only filter out disabled:true', () => {
        const data = {
            foo: 'bar',
            bar: { disabled: false },
            baz: { disabled: true }
        };
        const output = mapStateToProps(getInput(data));
        expect(output).toEqual(getOutput(data));
    });
});

describe('render ConnectedBody component', () => {
    it('should render correctly', () => {
        const wrapper = shallow(
            <Provider store={ store }>
                <ConnectedBody/>
            </Provider>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('render Body component', () => {
    it('should render correctly', () => {
        const wrapper = shallow(
            <Body technologies={ activeTechnologiesMock }/>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
