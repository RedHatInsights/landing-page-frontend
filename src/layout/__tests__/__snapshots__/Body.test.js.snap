// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Body component should render correctly 1`] = `
<Fragment>
  <Hero
    className="ins-c-hero__small"
    data-ouia-component-type="hero"
    subtitle="Discover purpose-built services exclusively for Red Hat subscribers"
    title="Scale the management and operations of your IT infrastructure"
  />
  <PageSection
    className="pf-m-fill ins-p-landing__content"
  >
    <Component
      activeKey={0}
      data-ouia-navigation="true"
      isSecondary={true}
      onSelect={[Function]}
    >
      <Tab
        data-ouia-component-id="nav-tab-1"
        data-ouia-component-type="tab"
        data-ouia-navigation-name="Tab 1"
        eventKey={0}
        key="0"
        title={
          <BannerCard
            category-id="1"
            description="this is a description"
            title="title"
          />
        }
      >
        <div
          className="ins-l-app-grid"
        >
          <FancyLink
            application-id="app-1"
            className="ins-l-app-grid__item"
            data-ouia-component-id="app-link-app-1"
            data-ouia-component-type="app-link"
            key="0"
            to="http://localhost/app-url-1"
          >
            app-name-1
          </FancyLink>
        </div>
      </Tab>
      <Tab
        data-ouia-component-id="nav-tab-2"
        data-ouia-component-type="tab"
        data-ouia-navigation-name="Tab 2"
        eventKey={1}
        key="1"
        title={
          <BannerCard
            category-id="2"
            description="this is a second description"
            title="secondtitle"
          />
        }
      >
        <div
          className="ins-l-app-grid"
        >
          <FancyLink
            application-id="app-2"
            className="ins-l-app-grid__item"
            data-ouia-component-id="app-link-app-2"
            data-ouia-component-type="app-link"
            key="0"
            to="http://localhost/app-url-2"
          >
            app-name-2
          </FancyLink>
        </div>
      </Tab>
    </Component>
  </PageSection>
</Fragment>
`;

exports[`render ConnectedBody component should render correctly 1`] = `<Connect(Body) />`;
